### ISSUE TITLE
[BUG] trap hook terminated

### ISSUE THREAD

**Issue by diachedelic** [npm/cli#6147]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

Scripts run via `npm run` have their trap hooks terminated prematurely. This is undesirable because the script does not get a chance to clean up after itself. Worse, it might clean up only some of its mess.

### Expected Behavior

Any Bash "trap" hooks to run to completion.

### Steps To Reproduce

Create a file demo.sh containing

```bash
#!/bin/bash
trap "echo start; sleep 1; echo end" EXIT
cat
```

and make it executable with `chmod +x demo.sh`. Notice that it has a trap hook that runs on exit. The hook echoes "start", does some work, then echoes "end". Declare it as the script "demo" in a package.json file:

```json
{
    "scripts":{
        "demo":"./demo.sh"
    }
}
```

Run the "demo" script, then interrupt it by typing ctrl-c. Notice that the trap hook does not run to completion:

```
$ npm run demo

> demo
> ./demo.sh

^Cstart
```

On the other hand, running the script directly runs the trap hook to completion:

```bash
$ ./demo.sh
^Cstart
end
```

### Environment

- npm: 9.4.0
- Node.js: v19.6.0
- OS Name: MacOS 13.1
- System Model Name: MacBook Pro
- npm config:
```ini
; "user" config from /Users/me/.npmrc

//registry.npmjs.org/:_authToken = (protected) 

; node bin location = /nix/store/fnhfi19p4wlj38wjrxipnb355jxf3zzn-nodejs-19.6.0/bin/node
; node version = v19.6.0
; npm local prefix = /private/tmp
; npm version = 9.4.0
; cwd = /private/tmp
; HOME = /Users/me
; Run `npm config ls -l` to show all defaults.
```

**Comment by fritzy** [comment]
I'm not able to reproduce this in `npm@9.6.0`. Please reopen if this is still an issue.

```
projects/scratchpad/test3 via ⬢ v19.0.1 
❯ npm -v          
9.6.0

projects/scratchpad/test3 via ⬢ v19.0.1 
❯ cat package.json
{
  "name": "test3",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "demo": "./test.sh"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}

projects/scratchpad/test3 via ⬢ v19.0.1 
❯ npm run demo

> test3@1.0.0 demo
> ./test.sh

^Cstart
end
```

**Comment by fritzy** [comment]
Looks like this can still happen in some cases.

**Comment by diachedelic** [comment]
Yes, I'm still seeing it with NPM v9.6.0.

```
$ node_modules/.bin/npm run demo

> demo
> ./demo.sh

^Cstart

$ node_modules/.bin/npm -v
9.6.0
```

**Comment by fritzy** [comment]
https://github.com/npm/run-script/pull/142

**Comment by diachedelic** [comment]
Seems to be fixed in NPM v9.7.2.
